### PR TITLE
Amit/claim functions

### PR DESCRIPTION
**Pull request by amit-shah-cb**

### CODE REVIEW

File: src/utils/claims/transactionCheck.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+export type CheckExectionParams = {
+  function: string;
+} & WebhookData;
+
+export async function checkFunctionExecution(
+  params: CheckExectionParams,
+  provider: ethers.JsonRpcProvider
+): Promise<boolean> {
+  const selector = ethers.id(params.function).substring(0, 10);
+  let tx = await provider.getTransaction(params.transaction_hash);
```
Comment (amit-shah-cb) [inline src/utils/claims/transactionCheck.ts:15]: /nit const

File: src/utils/claims/txHistoryCheck.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      params: [a, 'latest'],
+    };
+  }) as Array<{
+    id: number;
+    jsonrpc: '2.0';
+    method: string;
+    params: Array<any>;
+  }>;
+  const results = await provider._send(r);
+  const rr = results
```
Comment (amit-shah-cb) [inline src/utils/claims/txHistoryCheck.ts:100]: /nit change to `sum`

File: src/utils/claims/selectors.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  [CheckFunctionType.checkFunctionExecution]: checkFunctionExecution,
+  [CheckFunctionType.checkBalance]: checkBalance,
+  [CheckFunctionType.checkTokenIdBalance]: checkTokenIdBalance,
+  [CheckFunctionType.checkTxCountBatch]: checkTxCountBatch,
+};
+
+export const MapChallengeTypeUserAddress: {
+  [key in CheckFunctionType]: (w: any) => string | undefined;
+} = {
+  [CheckFunctionType.checkMint]: (w: WebhookData) => w.to_address,
```
Comment (amit-shah-cb) [inline src/utils/claims/selectors.ts:28]: toLowerCase()!

File: src/utils/claims/selectors.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    body: CheckBalanceParams
+  ): string {
+    return body.userAddress.toLowerCase();
+  },
+  [CheckFunctionType.checkTxCountBatch]: function (
+    body: CheckTxCountBatchParams
+  ): string | undefined {
+    return body.userAddress.toLowerCase();
+  },
+  [CheckFunctionType.checkTrivia]: function (o: any): string {
```
Comment (amit-shah-cb) [inline src/utils/claims/selectors.ts:41]: check how to not make this any

File: src/utils/claims/selectors.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    throw new Error('Function not implemented.');
+  },
+  [CheckFunctionType.checkTrivia]: function (body: object): boolean {
+    return body && body.hasOwnProperty('submitted');
+  },
+  [CheckFunctionType.checkFunctionExecution]: function (body: object): boolean {
+    throw new Error('Function not implemented.');
+  },
+  [CheckFunctionType.checkBalance]: function (body: object): boolean {
+    return true;
```
Comment (amit-shah-cb) [inline src/utils/claims/selectors.ts:87]: check if body.hasOwnProperty("userAddress")

File: src/utils/claims/selectors.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  },
+  [CheckFunctionType.checkTokenIdBalance]: function (body: object): boolean {
+    return true;
```
Comment (amit-shah-cb) [inline src/utils/claims/selectors.ts:90]: check if body.hasOwnProperty("userAddress")

File: package.json [hunk 7]
```diff
@@ -6,7 +6,8 @@
     "dev": "next dev",
```
Comment (amit-shah-cb) [inline package.json:6]: add `make gen-types`

File: package.json [hunk 8]
```diff
@@ -3,10 +3,11 @@
   "version": "0.1.0",
   "private": true,
   "scripts": {
-    "dev": "next dev",
+    "dev": "make gen-types && next dev",
```
Comment (siwan-cb) [inline package.json:6]: 👍

**Review by siwan-cb — APPROVED** [review]

**Pull request merged**